### ISSUE TITLE
check for existing output files disregards naming convention from cwl file

### ISSUE THREAD

**Issue by FelixErnst** [tgirke/systemPipeR#20]
Hi,

the existing result file check in `runCommandline` behaves not as expected. The `output_update` function reverts the file names to `SampleName` and disregard the naming convention from the `cwl` file. 

```
library(systemPipeR)
setwd("/home/felix/workflows/miCLIP/")
param_path <- "param/"
cwl_path <- paste0(param_path,"cwl/")
targets <- paste0(param_path,"targets.txt")
args_rtRNA <- loadWorkflow(targets = targets, wf_file = "bwa-pe-rtRNA.cwl", 
                           input_file = "bwa-pe-rtRNA.yml", dir_path = cwl_path)
args_rtRNA <- renderWF(args_rtRNA, inputvars = c(FileName1 = "_FASTQ_PATH1_", FileName2 = "_FASTQ_PATH2_", 
                                                 SampleName = "_SampleName_"))
output(args_rtRNA)[1]
#> $Wt1A
#> $Wt1A$`bwa-pe-rtRNA-miCLIP`
#> [1] "results/Wt1A-rtRNA.sam"
args_rtRNA_update <- output_update(args_rtRNA, dir=FALSE, replace=TRUE, extension=c(".sam", ".bam"))
output(args_rtRNA_update)[1
#> $Wt1A
#> $Wt1A$`bwa-pe-rtRNA-miCLIP`
#> [1] "/home/felix/workflows/miCLIP/results/Wt1A.bam"
```

As you can see the file names are created in the correct way during rendering, but the `output_update` function changes the filenames, which bounces the `file.exists` check, when such a file exists already.

Is this intended behavior?

Alternatively: How can I add a suffix/prefix to output names? I checked other example cwl files and similar things are done there, for example the `trimmomatic-pe.cwl` file in the example vignette:

```
outputs:
  trimmomatic_1_paired:
    label: "Paired forward reads"
    type: File
    outputBinding:
      glob: $(inputs.results_path.basename)/$(inputs.SampleName)_1P.trimmed.fastq.gz
```

Thanks for any advice.

Felix

[test.zip](https://github.com/tgirke/systemPipeR/files/4302113/test.zip)

<details>
<summary>sessionInfo</summary>
<pre>
#> R version 3.6.2 (2019-12-12)
#> Platform: x86_64-pc-linux-gnu (64-bit)
#> Running under: Debian GNU/Linux 10 (buster)
#> 
#> Matrix products: default
#> BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.3.5.so
#> 
#> locale:
#>  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
#>  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
#>  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
#>  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
#>  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
#> [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
#> 
#> attached base packages:
#> [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
#> [8] methods   base     
#> 
#> other attached packages:
#>  [1] systemPipeR_1.20.0          ShortRead_1.44.3           
#>  [3] GenomicAlignments_1.22.1    SummarizedExperiment_1.16.1
#>  [5] DelayedArray_0.12.2         matrixStats_0.55.0         
#>  [7] Biobase_2.46.0              BiocParallel_1.20.1        
#>  [9] Rsamtools_2.2.3             Biostrings_2.54.0          
#> [11] XVector_0.26.0              GenomicRanges_1.38.0       
#> [13] GenomeInfoDb_1.22.0         IRanges_2.20.2             
#> [15] S4Vectors_0.24.3            BiocGenerics_0.32.0        
#> 
#> loaded via a namespace (and not attached):
#>  [1] Category_2.52.1          bitops_1.0-6             bit64_0.9-7             
#>  [4] RColorBrewer_1.1-2       progress_1.2.2           httr_1.4.1              
#>  [7] Rgraphviz_2.30.0         backports_1.1.5          tools_3.6.2             
#> [10] R6_2.4.1                 DBI_1.1.0                colorspace_1.4-1        
#> [13] withr_2.1.2              tidyselect_1.0.0         prettyunits_1.1.1       
#> [16] bit_1.1-15.2             curl_4.3                 compiler_3.6.2          
#> [19] graph_1.64.0             rtracklayer_1.46.0       checkmate_2.0.0         
#> [22] scales_1.1.0             genefilter_1.68.0        RBGL_1.62.1             
#> [25] askpass_1.1              rappdirs_0.3.1           stringr_1.4.0           
#> [28] digest_0.6.25            rmarkdown_2.1            AnnotationForge_1.28.0  
#> [31] jpeg_0.1-8.1             pkgconfig_2.0.3          htmltools_0.4.0         
#> [34] BSgenome_1.54.0          dbplyr_1.4.2             limma_3.42.2            
#> [37] highr_0.8                rlang_0.4.5              RSQLite_2.2.0           
#> [40] GOstats_2.52.0           hwriter_1.3.2            dplyr_0.8.4             
#> [43] VariantAnnotation_1.32.0 RCurl_1.98-1.1           magrittr_1.5            
#> [46] GO.db_3.10.0             GenomeInfoDbData_1.2.2   Matrix_1.2-18           
#> [49] Rcpp_1.0.3               munsell_0.5.0            lifecycle_0.1.0         
#> [52] stringi_1.4.6            yaml_2.2.1               edgeR_3.28.1            
#> [55] zlibbioc_1.32.0          BiocFileCache_1.10.2     grid_3.6.2              
#> [58] blob_1.2.1               crayon_1.3.4             lattice_0.20-40         
#> [61] splines_3.6.2            GenomicFeatures_1.38.2   annotate_1.64.0         
#> [64] hms_0.5.3                batchtools_0.9.12        locfit_1.5-9.1          
#> [67] knitr_1.28               pillar_1.4.3             rjson_0.2.20            
#> [70] base64url_1.4            biomaRt_2.42.0           XML_3.99-0.3            
#> [73] glue_1.3.1               evaluate_0.14            latticeExtra_0.6-29     
#> [76] data.table_1.12.8        png_0.1-7                vctrs_0.2.3             
#> [79] gtable_0.3.0             openssl_1.4.1            purrr_0.3.3             
#> [82] assertthat_0.2.1         ggplot2_3.3.0            xfun_0.12               
#> [85] xtable_1.8-4             survival_3.1-8           pheatmap_1.0.12         
#> [88] tibble_2.1.3             AnnotationDbi_1.48.0     memoise_1.1.0           
#> [91] brew_1.0-6               GSEABase_1.48.0
</pre>
</details>

**Comment by dcassol** [comment]
Hi @FelixErnst, 
The ```output_update``` function was design to update the output on the args object in one specific situation: when you use ```clusterRun``` function (which internally is running ```runCommandline```) to submit the jobs in the cluster, and also it was set ```make_bam``` is true.
```
reg <- clusterRun(args, FUN = runCommandline, more.args = list(args = args, 
    make_bam = TRUE, dir = FALSE), conffile = ".batchtools.conf.R", 
    template = "batchtools.slurm.tmpl", Njobs = 18, runid = "01", 
    resourceList = resources)
args <- output_update(args, dir = FALSE, replace = TRUE, extension = c(".sam", 
    ".bam"))  ## Updates the output(args) to the right location in the subfolders
output(args)
```
In the cases you set ```make_bam=FALSE```, you should not use  ```output_update``` because the information on the slot output are correct. 

Another detail is when you are using ```runCommadline``` as follows, and ```make_bam=TRUE```, the function internally already updates the output, replacing *sam for *bam and assign to the object ```args``` in the example:
```
args <- runCommandline(args, make_bam=TRUE)
```
To answer your second question, yes, you can definitely add a suffix to the output files. In this case, you need to modify the ```*.cwl``` file it`s used on the ```loadWorkflow``` function and add the respective information as in the example you showed here.

I hope I answered your questions. If you need any help to custom your ```cwl files``` let me know. 

Thank you!
Dani

**Comment by FelixErnst** [comment]
Hi @dcassol,

Thanks for getting into the meaning of the `output_update`. I used the `output_update` function in my example above because, that is where my problem comes from. Sorry for not going into details enough. Here it goes:

I want to run two seperate mapping steps on the same input. The first one, the default BWA PE mapping strategy works fine and produces a bam file, e.g. `Wt1A.bam`. Running the second analysis with the `cwl` provided here, `runCommandline` reports output file present and does nothing.

The problem is located here https://github.com/tgirke/systemPipeR/blob/e09213184d76570d919986821bef9f0781a3bc8e/R/utilities.R#L182 and the lines following that. https://github.com/tgirke/systemPipeR/blob/e09213184d76570d919986821bef9f0781a3bc8e/R/utilities.R#L187 checks for the `file.exists`. Because the original workflow succeeded and `args` gets updated wrongly, the file is present stopping my workflow.

My guess would be that it is not intendend. Is that correct?

Felix

**Comment by dcassol** [comment]
Hi @FelixErnst, 
Thank you for letting me know about this bug. I will fix the bug and let you know; this was not supposed to happen.

Thank you very much, 
Dani

**Comment by FelixErnst** [comment]
Hi @dcassol,

I got another issue regarding output path. Can you take a look at it?

When I change the output directory to include an additional subfolder, the `output()` includes only the last folder.

So in the original `.yml` I changed the following part to
```
results_path:
  class: Directory
  path: ./results
```

this
```
results_path:
  class: Directory
  path: ./results/tx
```

This results in the following output. (Please not the different version of the `.yml` file)
```
param_path <- "param/"
cwl_path <- paste0(param_path,"cwl/")
targets <- paste0(param_path,"targets.txt")
args_rtRNA <- loadWorkflow(targets = targets, wf_file = "bwa-pe-rtRNA.cwl", 
                               input_file = "bwa-pe-rtRNA.yml", dir_path = cwl_path)
args_rtRNA <- renderWF(args_rtRNA, inputvars = c(FileName1 = "_FASTQ_PATH1_", FileName2 = "_FASTQ_PATH2_", 
                                                     SampleName = "_SampleName_"))
output(args_rtRNA)[1]
#> $Wt1A
#> $Wt1A$`bwa-pe-rtRNA-miCLIP`
#> [1] "results/Wt1A-rtRNA.sam"
args_rtRNA <- loadWorkflow(targets = targets, wf_file = "bwa-pe-rtRNA.cwl", 
                               input_file = "bwa-pe-rtRNA2.yml", dir_path = cwl_path)
args_rtRNA <- renderWF(args_rtRNA, inputvars = c(FileName1 = "_FASTQ_PATH1_", FileName2 = "_FASTQ_PATH2_", 
                                                     SampleName = "_SampleName_"))
output(args_rtRNA)[1]
#> $Wt1A
#> $Wt1A$`bwa-pe-rtRNA-miCLIP`
#> [1] "tx/Wt1A-rtRNA.sam"
```

Are nested subfolders not supported?

Thanks for any advice and looking into it.

Felix

**Comment by dcassol** [comment]
Hi @FelixErnst, 
Nested subfolders are supported; I will investigate and see what is happening. 
One quick way to do precisely what you want to accomplish is:
```
param_path <- "param/"
cwl_path <- paste0(param_path,"cwl/")
targets <- paste0(param_path,"targets.txt")
args <- loadWorkflow(targets = targets, wf_file = "bwa-pe.cwl", 
                               input_file = "bwa-pe.yml", dir_path = cwl_path)
args <- renderWF(args, inputvars = c(FileName1 = "_FASTQ_PATH1_", FileName2 = "_FASTQ_PATH2_", SampleName = "_SampleName_"))
output(args)[1]
#> $Wt1A
#> $Wt1A$`bwa-pe`
#> [1] "results/Wt1A.sam"
args <- runCommandline(args, make_bam=TRUE, dir=TRUE)
output(args)[1]
#> $Wt1A
#> $Wt1A$`bwa-pe`
#> [1] "results/bwa-pe/Wt1A/Wt1A.sam"
````
````
args_rtRNA <- loadWorkflow(targets = targets, wf_file = "bwa-pe-rtRNA.cwl",  input_file = "bwa-pe-rtRNA.yml", dir_path = cwl_path)
args_rtRNA <- renderWF(args_rtRNA, inputvars = c(FileName1 = "_FASTQ_PATH1_", FileName2 = "_FASTQ_PATH2_",  SampleName = "_SampleName_"))
output(args_rtRNA)[1]
#> $Wt1A
#> $Wt1A$`bwa-pe-rtRNA`
#> [1] "results/Wt1A-rtRNA.sam"
args <- runCommandline(args, make_bam=TRUE, dir=TRUE)
output(args)[1]
#> $Wt1A
#> $Wt1A$`bwa-pe-rtRNA`
#> [1] "results/wa-pe-rtRNA/Wt1A/Wt1A.sam"
````
```runCommandlie``` allows to create subfolder for all the different softawes/analysis and also for each one of the samples. In my opinion, this is a better way to organize the results.

**Comment by dcassol** [comment]
@FelixErnst, in the meantime, I am still working on the bugs you reported, and I will let you know when this is fixed. Thank you very much!

**Comment by FelixErnst** [comment]
@dcassol Thanks for working on this! The option with `runCommandline` is a good alternative. I had read it in the man pages, but didn't keep track of it.

I also checked whether the `basename` function in the `cwl` file does have any influence on the output folder, but apparently it doesn't.

**Comment by dcassol** [comment]
@FelixErnst, I have been working on ```runCommandline```, creating a more robust and modular function. I have committed a fix for it on GitHub and Bioc (Version 1.21.6). 
Please let me know if it is working for you. 
I am going to close this for now, but feel free to re-open if you have more questions. 
Thanks for reporting it.
Cheers, 
Dani
